lib/Optimizer/Dialect/FIROps.cpp

```diff
@@ -1045,6 +1045,9 @@ mlir::FuncOp fir::createFuncOp(mlir::Location loc, mlir::ModuleOp module,
   if (auto f = module.lookupSymbol<mlir::FuncOp>(name))
     return f;
   mlir::OpBuilder modBuilder(module.getBodyRegion());
+  // Insert new function at the end of the current module.
+  assert(modBuilder.getInsertionBlock() && "builder must have insertion block");
```

Comment:
Do we need this assert? Is the insertion point guaranteed from the constructor above?

Also: this must likely should be a FirOpBuilder, not the base class `mlir::OpBuilder`. I must've missed it in the last cleanup.

Reply:
Probably,  I am OK to remove the assert if you are sure about this, I usually always assert before a `->` just because. I have not moved it as I am not 100% sure what you have in mind regarding moving this in `FirOpBuilder`.

Reply:
I believe that if we are passing in a valid `ModuleOp`, then the insertion point will always be set.

- Creating a `ModuleOp` results in calls to `addRegion()` and `ensureTerminator()`. These should create the Module -> Region -> Block structure. Also, since it's adding the module terminator, the block is not empty.
- The constructor of the `OpBuilder` should then call `setInsertionPoint()`.

We can leave the assertion in for now, but maybe add a `// FIXME` that we suspect we're over asserting here.

Reply:
The comment about FirOpBuilder makes no sense. I was confused.